cmd/azureBlobUpload.go

```diff
@@ -0,0 +1,119 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+)
+
+// AzureContainerAPI is used to mock Azure containerClients in unit tests
+type AzureContainerAPI interface {
+	NewBlockBlobClient(blobName string) azblob.BlockBlobClient
+}
+
+// NewBlockBlobClient creates a blockBlobClient from a containerClient
+func NewBlockBlobClient(blobName string, api AzureContainerAPI) azblob.BlockBlobClient {
+	return api.NewBlockBlobClient(blobName)
+}
+
+// UploadFile uploads a file to an Azure Blob Storage
+// The function is uses the UploadFileToBlockBlob function from the Azure SDK
+// We introduce this 'wrapper' for mocking reasons
+func UploadFile(ctx context.Context, api *azblob.BlockBlobClient, file *os.File, o azblob.HighLevelUploadToBlockBlobOption) (*http.Response, error) {
+	return api.UploadFileToBlockBlob(ctx, file, o)
+}
+
+// Struct to store Azure credentials from specified JSON string
+type azureCredentials struct {
+	SASToken    string `json:"sas_token"`
+	AccountName string `json:"account_name"`
+	Container   string `json:"container_name"`
+}
+
+func azureBlobUpload(config azureBlobUploadOptions, telemetryData *telemetry.CustomData) {
+	// Prepare Credentials
+	log.Entry().Infoln("Start reading Azure Credentials")
+	var creds azureCredentials
+
+	err := json.Unmarshal([]byte(config.JSONCredentialsAzure), &creds)
+	if err != nil {
+		log.Entry().
+			WithError(err).
+			Fatal("Could not read JSONCredentialsAzure")
+	}
+
+	// Initialize Azure Service Client
+	sasURL := fmt.Sprintf("https://%s.blob.core.windows.net/?%s", creds.AccountName, creds.SASToken)
+	serviceClient, err := azblob.NewServiceClientWithNoCredential(sasURL, nil)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Could not instantiate Azure Service Client!")
+	}
+
+	// Get a containerClient from ServiceClient
+	containerClient := serviceClient.NewContainerClient(creds.Container)
```

Comment:
How about adding this preparation to the beginning of `runAzureBlobUpload`, which in turn returns an error if something goes wrong. Then there is only one call to `log.Entry().WithError(err).Fatal` necessary.
The reason we do this is to be able to test more code. The `azureBlobUpload` is usually not tested, which is why we keep that code to a minimum.

Reply:
If we put the initialisation of the Azure-Clients into runAzureBlobUpload, how would you mock these for the unit tests?

Reply:
I did not think of that. What you can do is split it into two functions that are called by `runAzureBlobUpload`, as in:
```
func runAzureBlobUpload(...) error {
    x, y := init()
    return uploadFiles(x, y, z)
}
```

Reply:
I have now implemented this (hopefully as intended)... However, in the end, the error handling has only moved from one function to the other.

Reply:
By adding an `init` (or `setup`) function as well you could add unit tests for that part as well. Up to you to decide if that is necessary. I for one like to test that the error messages are as expected and have the idea that no-body will accidentally do changes to the setup, for example changing the url to SAS, etc.

Reply:
A setup function has now been implemented and is already being used to check the credentials with "validator" in the tests. 

However, checking the Azure clients and thus also the SAS URL is not so easy here: Although the functions for creating the clients (e.g.: NewServiceClientWithNoCredential()) have error as the return type, these returned errors are always "nil". At the moment, these are only integrated in the SDK as placeholders. My guess is that this will change with future releases. 

The SAS URL is therefore only cached when the client is created and then only used during the actual upload.
If you want to check this yourself, you can find the corresponding code of the SDK here: https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob